cpp/include/cudf/datetime.hpp

```diff
@@ -378,5 +378,96 @@ std::unique_ptr<column> ceil_nanosecond(
   column_view const& column,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief Round down to the nearest day
+ *
+ * @param column cudf::column_view of the input datetime values
+ * @param mr Device memory resource used to allocate device memory of the returned column.
+ *
+ * @throw cudf::logic_error if input column datatype is not TIMESTAMP
+ * @return cudf::column of the same datetime resolution as the input column
+ */
+std::unique_ptr<cudf::column> floor_day(
```

Comment:
Just a thought (should not block this PR) - can we use a enum type to specify which component to round for the timestamp instead of having one function for each?

Reply:
It turns out that you're not the first to suggest this, @isVoid. 😄 https://github.com/rapidsai/cudf/pull/8942/files#r691691807

It seems like we have multiple current/past PR reviewers aligned on the idea that an enum would be better than multiple functions. I'm meeting with @skirui-source @mayankanand007 about this PR now, we will update this PR with our discussion outcomes soon.

Reply:
There was also a brief Slack thread about this that I can share with you offline.

Reply:
Not mean to make the water muddier but would this be a good opportunity to test out [templated public API](https://github.com/rapidsai/cudf/issues/6495)?

Reply:
@isVoid could you elaborate a bit more on how [templated public API](https://github.com/rapidsai/cudf/issues/6495) is related to this issue? Thanks!

Reply:
I think that's probably a much bigger question that we'd want to make a plan for libcudf as a whole and prototype with existing well-defined APIs. Datetime APIs like this might be good guinea pigs, but I wouldn't bother thinking about that until this PR is merged as is.